File: src/familiar_connect/tts_player/logging_player.py
"""Logging-only TTS player.

Phase-2 production default — logs what it *would* speak but does not
produce audio. Discord voice-channel playback is a follow-up phase;
until then the responder's output is visible in the log stream so the
pipeline can be exercised end-to-end without hitting a Cartesia/Azure
API.
"""

from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls

if TYPE_CHECKING:
    from familiar_connect.bus.envelope import TurnScope

_logger = logging.getLogger("familiar_connect.tts_player.logging")


class LoggingTTSPlayer:
    """No-audio TTS stand-in.

    Honours cancellation semantics so the barge-in loop behaves the
    same as with a real player.
    """

    def __init__(self, *, ms_per_word: int = 200, poll_ms: int = 20) -> None:
        self._ms_per_word = ms_per_word
        self._poll_ms = poll_ms
        self._stop_event = asyncio.Event()

    async def speak(self, text: str, *, scope: TurnScope) -> None:
        words = text.split()
        budget_ms = len(words) * self._ms_per_word
        played_ms = 0
        _logger.info(
            f"{ls.tag('🔊 Say', ls.G)} "
            f"{ls.kv('turn', scope.turn_id, vc=ls.LC)} "
            f"{ls.kv('words', str(len(words)), vc=ls.LY)} "
            f"{ls.kv('text', ls.trunc(text, 200), vc=ls.LW)}"
        )
        self._stop_event = asyncio.Event()

        while played_ms < budget_ms:
            if scope.is_cancelled() or self._stop_event.is_set():
                _logger.info(
                    f"{ls.tag('🔊 Cut', ls.Y)} "
                    f"{ls.kv('turn', scope.turn_id, vc=ls.LC)} "
                    f"{ls.kv('played_ms', str(played_ms), vc=ls.LY)}"
                )
                return
            step = min(self._poll_ms, budget_ms - played_ms)
            try:
                await asyncio.wait_for(
                    self._stop_event.wait(),
                    timeout=step / 1000.0,
                )
            except TimeoutError:
                played_ms += step
                continue
            return

    async def stop(self) -> None:
        self._stop_event.set()

File: rust/familiar-connect/src/tts_player/logging_player.rs
//! Logging-only TTS player (subsystem 09; Python `tts_player/logging_player.py`).
//!
//! Phase-2 production default: logs what it *would* speak but produces no audio.
//! Honours cancellation so the barge-in loop behaves the same as with a real
//! player. Pacing sleeps `ms_per_word` per whitespace word, polling
//! cancellation / stop every `poll_ms`.

use std::sync::Mutex;
use std::time::Duration;

use async_trait::async_trait;
use tokio_util::sync::CancellationToken;

use crate::bus::envelope::TurnScope;
use crate::log_style as ls;
use crate::support::text::truncate;
use crate::tts_player::protocol::TtsPlayer;

/// No-audio TTS stand-in that paces + honours cancellation.
pub struct LoggingTTSPlayer {
    ms_per_word: u64,
    poll_ms: u64,
    /// Re-created per `speak`; `stop` cancels the current utterance only.
    stop_token: Mutex<CancellationToken>,
}

impl LoggingTTSPlayer {
    /// Player with the given pacing (`ms_per_word`) and cancel granularity
    /// (`poll_ms`).
    #[must_use]
    pub fn new(ms_per_word: u64, poll_ms: u64) -> Self {
        Self {
            ms_per_word,
            poll_ms,
            stop_token: Mutex::new(CancellationToken::new()),
        }
    }
}

impl Default for LoggingTTSPlayer {
    fn default() -> Self {
        Self::new(200, 20)
    }
}

#[async_trait]
impl TtsPlayer for LoggingTTSPlayer {
    async fn speak(&self, text: &str, scope: &TurnScope) {
        let words = text.split_whitespace().count() as u64;
        let budget_ms = words * self.ms_per_word;
        let mut played_ms: u64 = 0;

        tracing::info!(
            target: "familiar_connect.tts_player.logging",
            "{} {} {} {}",
            ls::tag("🔊 Say", ls::G),
            ls::kv_styled("turn", &scope.turn_id, ls::W, ls::LC),
            ls::kv_styled("words", &words.to_string(), ls::W, ls::LY),
            ls::kv_styled("text", &truncate(text, 200), ls::W, ls::LW),
        );

        // Reset the per-call stop gate on re-entry.
        let token = CancellationToken::new();
        *self.stop_token.lock().expect("stop_token poisoned") = token.clone();

        while played_ms < budget_ms {
            if scope.is_cancelled() || token.is_cancelled() {
                tracing::info!(
                    target: "familiar_connect.tts_player.logging",
                    "{} {} {}",
                    ls::tag("🔊 Cut", ls::Y),
                    ls::kv_styled("turn", &scope.turn_id, ls::W, ls::LC),
                    ls::kv_styled("played_ms", &played_ms.to_string(), ls::W, ls::LY),
                );
                return;
            }
            let step = self.poll_ms.min(budget_ms - played_ms);
            tokio::select! {
                () = token.cancelled() => return,
                () = tokio::time::sleep(Duration::from_millis(step)) => played_ms += step,
            }
        }
    }

    async fn stop(&self) {
        self.stop_token
            .lock()
            .expect("stop_token poisoned")
            .cancel();
    }
}
